iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -4,1310 +4,1097 @@
-            _deviceboundMessagePrefix = string.Format(CultureInfo.InvariantCulture, DeviceBoundMessagesTopicPrefix, connectionCredentials.DeviceId);
+                    if (proxy.Credentials != null)
+                    {
+                        NetworkCredential credentials = proxy.Credentials.GetCredential(serviceUri, "Basic");
+                        string username = credentials.UserName;
+                        string password = credentials.Password;
+                        _mqttClientOptionsBuilder.WithProxy(proxyUri.AbsoluteUri, username, password);
+                    }
+                    else
+                    {
+                        _mqttClientOptionsBuilder.WithProxy(proxyUri.AbsoluteUri);
+                    }
+                }
+            }
+            else
+            {
+                // "ssl://" prefix is not needed here
+                var uri = _hostName;
+                _mqttClientOptionsBuilder.WithTcpServer(uri, ProtocolGatewayPort);
+            }
 
-            _qosSendPacketToService = settings.PublishToServerQoS;
-            _qosReceivePacketFromService = settings.ReceivingQoS;
+            MqttClientOptionsBuilderTlsParameters tlsParameters = new MqttClientOptionsBuilderTlsParameters();
 
-            // If the CleanSession flag is set to false, C2D messages will be retained across device sessions, i.e. the device
-            // will receive the C2D messages that were sent to it while it was disconnected.
-            // If the CleanSession flag is set to true, the device will receive only those C2D messages that were sent
-            // after it had subscribed to the message topic.
-            _retainMessagesAcrossSessions = !settings.CleanSession;
+            List<X509Certificate> certs = _connectionCredentials.Certificate == null
+                ? new List<X509Certificate>(0)
+                : new List<X509Certificate> { _connectionCredentials.Certificate };
 
-            _webProxy = settings.Proxy;
+            tlsParameters.Certificates = certs;
 
-            if (channelFactory != null)
+            if (_mqttTransportSettings?.RemoteCertificateValidationCallback != null)
             {
-                _channelFactory = channelFactory;
+                tlsParameters.CertificateValidationHandler = certificateValidationHandler;
             }
-            else
+
+            tlsParameters.UseTls = true;
+            tlsParameters.SslProtocol = _mqttTransportSettings.SslProtocols;
+            _mqttClientOptionsBuilder.WithTls(tlsParameters);
+
+            _mqttClientOptionsBuilder.WithProtocolVersion(MQTTnet.Formatter.MqttProtocolVersion.V311); // 3.1.1
+
+            _mqttClientOptionsBuilder.WithCleanSession(_mqttTransportSettings.CleanSession);
+
+            _mqttClientOptionsBuilder.WithKeepAlivePeriod(_mqttTransportSettings.KeepAlive);
```

Comment:
You can chain these, IIRC.

```suggestion
            _mqttClientOptionsBuilder
                .WithTls(tlsParameters)
                .WithProtocolVersion(MQTTnet.Formatter.MqttProtocolVersion.V311) // 3.1.1
                .WithCleanSession(_mqttTransportSettings.CleanSession)
                .WithKeepAlivePeriod(_mqttTransportSettings.KeepAlive);
```